### ISSUE TITLE
RegexNER Annotator: Whitespace inside capture groups

### ISSUE THREAD

**Issue by kopperud** [stanfordnlp/CoreNLP#746]
Hello,

I am having some trouble with a regex expression that I want to match as a named-entity:

```regex
([A-Z][a-z]+ )+(Dog|Cat)
```
Here is some sample text

> The little Brown Dog went into the forest. John had an old Egyptian Cat, and its fur was patchy.

Using the Regexr.com debugging tool, it works perfectly (sorry if the link is not persistent):

[https://regexr.com/3slph](https://regexr.com/3slph)

## In CoreNLP

Here is my regexner.mapping file
```
([A-Z][a-z]+ )+(Dog|Cat)     ANIMAL   ORGANIZATION,PERSON,LOCATION,MISC,NUMBER,MONEY,PERCENT,DATE,TIME        1.0
```

When I load the annotator and try the sentence in the debugging interface, I get this error:
```java
java.util.regex.PatternSyntaxException: Unclosed group near index 12 ([A-Z][a-z]+ ^
```

Here is the console print for the CoreNLP server:

```
[main] INFO CoreNLP - --- StanfordCoreNLPServer#main() called ---
[main] INFO CoreNLP - setting default constituency parser
[main] INFO CoreNLP - warning: cannot find edu/stanford/nlp/models/srparser/englishSR.ser.gz
[main] INFO CoreNLP - using: edu/stanford/nlp/models/lexparser/englishPCFG.ser.gz instead
[main] INFO CoreNLP - to use shift reduce parser download English models jar from:
[main] INFO CoreNLP - http://stanfordnlp.github.io/CoreNLP/download.html
[main] INFO CoreNLP -     Threads: 32
[main] INFO CoreNLP - Starting server...
[main] INFO CoreNLP - StanfordCoreNLPServer listening at /0:0:0:0:0:0:0:0:9000
[pool-1-thread-1] ERROR CoreNLP - Failure to load language specific properties: StanfordCoreNLP.properties for en
[pool-1-thread-1] INFO CoreNLP - [/0:0:0:0:0:0:0:1:57510] API call w/annotators tokenize,ssplit,pos,lemma,ner,regexner
The little Brown Dog went into the forest. John had an old Egyptian Cat, and its fur was patchy.
[pool-1-thread-1] INFO edu.stanford.nlp.pipeline.StanfordCoreNLP - Adding annotator tokenize
[pool-1-thread-1] INFO edu.stanford.nlp.pipeline.TokenizerAnnotator - No tokenizer type provided. Defaulting to PTBTokenizer.
[pool-1-thread-1] INFO edu.stanford.nlp.pipeline.StanfordCoreNLP - Adding annotator ssplit
[pool-1-thread-1] INFO edu.stanford.nlp.pipeline.StanfordCoreNLP - Adding annotator pos
[pool-1-thread-1] INFO edu.stanford.nlp.tagger.maxent.MaxentTagger - Loading POS tagger from edu/stanford/nlp/models/pos-tagger/english-left3words/english-left3words-distsim.tagger ... done [0.6 sec].
[pool-1-thread-1] INFO edu.stanford.nlp.pipeline.StanfordCoreNLP - Adding annotator lemma
[pool-1-thread-1] INFO edu.stanford.nlp.pipeline.StanfordCoreNLP - Adding annotator ner
[pool-1-thread-1] INFO edu.stanford.nlp.ie.AbstractSequenceClassifier - Loading classifier from edu/stanford/nlp/models/ner/english.all.3class.distsim.crf.ser.gz ... done [0.9 sec].
[pool-1-thread-1] INFO edu.stanford.nlp.time.JollyDayHolidays - Initializing JollyDayHoliday for SUTime from classpath edu/stanford/nlp/models/sutime/jollyday/Holidays_sutime.xml as sutime.binder.1.
[pool-1-thread-1] INFO edu.stanford.nlp.time.TimeExpressionExtractorImpl - Using following SUTime rules: edu/stanford/nlp/models/sutime/defs.sutime.txt,edu/stanford/nlp/models/sutime/english.sutime.txt,edu/stanford/nlp/models/sutime/english.holidays.sutime.txt
[pool-1-thread-1] INFO edu.stanford.nlp.pipeline.TokensRegexNERAnnotator - TokensRegexNERAnnotator ner.fine.regexner: Read 580729 unique entries out of 581888 from edu/stanford/nlp/models/kbp/regexner_caseless.tab, 0 TokensRegex patterns.
[pool-1-thread-1] INFO edu.stanford.nlp.pipeline.TokensRegexNERAnnotator - TokensRegexNERAnnotator ner.fine.regexner: Read 4857 unique entries out of 4868 from edu/stanford/nlp/models/kbp/regexner_cased.tab, 0 TokensRegex patterns.
[pool-1-thread-1] INFO edu.stanford.nlp.pipeline.TokensRegexNERAnnotator - TokensRegexNERAnnotator ner.fine.regexner: Read 585586 unique entries from 2 files
[pool-1-thread-1] INFO edu.stanford.nlp.pipeline.StanfordCoreNLP - Adding annotator regexner
[pool-1-thread-1] INFO edu.stanford.nlp.pipeline.TokensRegexNERAnnotator - TokensRegexNERAnnotator regexner: Read 1 unique entries out of 1 from strat_rules.txt, 0 TokensRegex patterns.
java.util.regex.PatternSyntaxException: Unclosed group near index 12
([A-Z][a-z]+
            ^
        at java.util.regex.Pattern.error(Pattern.java:1957)
        at java.util.regex.Pattern.accept(Pattern.java:1815)
        at java.util.regex.Pattern.group0(Pattern.java:2910)
        at java.util.regex.Pattern.sequence(Pattern.java:2053)
        at java.util.regex.Pattern.expr(Pattern.java:1998)
        at java.util.regex.Pattern.compile(Pattern.java:1698)
        at java.util.regex.Pattern.<init>(Pattern.java:1351)
        at java.util.regex.Pattern.compile(Pattern.java:1054)
        at edu.stanford.nlp.ling.tokensregex.ComplexNodePattern$StringAnnotationRegexPattern.<init>(ComplexNodePattern.java:254)
        at edu.stanford.nlp.ling.tokensregex.ComplexNodePattern.newStringRegexPattern(ComplexNodePattern.java:53)
        at edu.stanford.nlp.ling.tokensregex.CoreMapNodePattern.valueOf(CoreMapNodePattern.java:45)
        at edu.stanford.nlp.pipeline.TokensRegexNERAnnotator.createPatternMatcher(TokensRegexNERAnnotator.java:355)
        at edu.stanford.nlp.pipeline.TokensRegexNERAnnotator.<init>(TokensRegexNERAnnotator.java:295)
        at edu.stanford.nlp.pipeline.AnnotatorImplementations.tokensRegexNER(AnnotatorImplementations.java:85)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.lambda$getNamedAnnotators$7(StanfordCoreNLP.java:548)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.lambda$null$30(StanfordCoreNLP.java:625)
        at edu.stanford.nlp.util.Lazy$3.compute(Lazy.java:126)
        at edu.stanford.nlp.util.Lazy.get(Lazy.java:31)
        at edu.stanford.nlp.pipeline.AnnotatorPool.get(AnnotatorPool.java:149)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.construct(StanfordCoreNLP.java:495)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.<init>(StanfordCoreNLP.java:201)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.<init>(StanfordCoreNLP.java:194)
        at edu.stanford.nlp.pipeline.StanfordCoreNLP.<init>(StanfordCoreNLP.java:181)
        at edu.stanford.nlp.pipeline.StanfordCoreNLPServer.mkStanfordCoreNLP(StanfordCoreNLPServer.java:345)
        at edu.stanford.nlp.pipeline.StanfordCoreNLPServer.access$700(StanfordCoreNLPServer.java:54)
        at edu.stanford.nlp.pipeline.StanfordCoreNLPServer$CoreNLPHandler.handle(StanfordCoreNLPServer.java:751)
        at com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:79)
        at sun.net.httpserver.AuthFilter.doFilter(AuthFilter.java:83)
        at com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:82)
        at sun.net.httpserver.ServerImpl$Exchange$LinkHandler.handle(ServerImpl.java:675)
        at com.sun.net.httpserver.Filter$Chain.doFilter(Filter.java:79)
        at sun.net.httpserver.ServerImpl$Exchange.run(ServerImpl.java:647)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

## Swap whitespace ` `for regex special whitespace character `\s`

I also tried this pattern
```
([A-Z][a-z]+\s)+(Dog|Cat)     ANIMAL   ORGANIZATION,PERSON,LOCATION,MISC,NUMBER,MONEY,PERCENT,DATE,TIME        1.0
```

With this, there are no errors, however my pattern does not get matched.

Any suggestions? What did I do wrong?

Best regards,
Bjørn Tore Kopperud

**Comment by J38** [comment]
Hi I have 2 suggestions.

So for one, RegexNER match over a sequence of tokens.  Each entry in the rule (separated by space) matches one token.

So this rule:

```
[A-Z][a-z]+ (Dog|Cat)
```

will at most match 2 tokens.  One capitalized word followed by Dog or Cat.  If you want to use this simpler rule format, you can just add a separate rule for 3 words followed by Dog or Cat, 2 words, etc...This might work if you at most need 4 words preceding Cat or Dog.

I would understand if you found it unsatisfying to have to have 4 (or more) versions of the same rule to handle each case.  If you want to avoid things like that, you will have to become familiar with the TokensRegex rules in general.  There is a very thorough tutorial here:

https://stanfordnlp.github.io/CoreNLP/tokensregex.html

Also, I will add some details to that tutorial about running the `tokensregex` annotator in a pipeline.  I would recommend reading this tutorial. 

Please let me know if you have any other questions.

**Comment by kopperud** [comment]
Hi J38, thank you for your advice.

I managed to solve the problem with this TokensRegex:

```
( [{ word:/[A-Z][a-z]+/ }]+? /Dog|Cat/ )
```

It solves my problem. I'm familiar with standard Regex, but understanding how this annotator is implemented on top of the TokensRegex is more challenging. 

I have one comment on the documentation of how to use the TokensRegex syntax:

(from bottom https://stanfordnlp.github.io/CoreNLP/regexner.html):

    a TokensRegex expression (marked by starting with “( “ and ending with “ )”. 
    Example: ( /University/ /of/ [ {ner:LOCATION} ] )\tSCHOOL

It is somewhat hidden and cryptic, I did not immediately understand I had to use whitespace after `(` and before `)`. Perhaps it is because of the font?

> Also, I will add some details to that tutorial about running the tokensregex annotator in a pipeline

I think this would be a great addition.



Thank you for your quick reply, in the end it works perfectly!
Bjørn
